### DISCUSSION TITLE
Generic types with `HelperLike`

### DISCUSSION THREAD

**Discussion by swastik** [typed-ember/glint#542] [Q&A] [answered]
Thanks for your work on this! 😄 I'm trying to type some external addons that provide helpers. For example:

```ts
import { HelperLike } from '@glint/template';

declare module '@glint/environment-ember-loose/registry' {
  export default interface Registry {
    or: HelperLike<{
      Args: { Positional: [any, any] };
      Return: any;
    }>;
  }
}
```

The helper can be invoked like this and it returns either of the two arguments (for example):

```hbs
{{#let (or a b) as |foo|}}
  {{foo}}
{{/let}}
```

However, this seems like it will return an incorrect type: `any`. Is there a way to type these correctly? Thanks for any help!

**Comment by chriskrycho** [comment, marked as answer]
Good question! I would write this roughly like this (off the top of my head, so I cannot *promise* this is exactly right, but it's directionally correct):

```ts
declare function or<T extends any[]>(...params: T): T[number];

declare module '@glint/environment-ember-loose/registry' {
  export default interface Registry {
    or: typeof or;
  }
}
```

Or, with a class-based helper:

```ts
import type Helper from '@ember/component/helper';

declare class OrHelper<T extends any[]> extends Helper<{
  Args: {
    Positional: T;
  };
  Return: T[number];
}> {
  compute(positional: T): T[number];
}

declare module '@glint/environment-ember-loose/registry' {
  export default interface Registry {
    or: typeof OrHelper;
  }
}
```

The use of the `T extends any[]`, `...params: T`, and `T[number]` is to use TypeScript's inference over tuple spread types so that what you're saying is: the positional params are some list of items, and the resulting type is the type of one of those items. You can see that working in [this playground][p].

[p]: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUwB5mQYxgFQJ4YDyAZnAL5zFQQhwDkaG2AtDAcnQNwBQ3AJlgA2AQyjJKAVwB22YBClxoAHhwpUMZFL4BnOMKl4A2gF0AfAAoAdNbCjhIbQC44OAJTOchqRJAAjZFDGPNyMWLjsJObQ5gBEwjEANHAxvjGurpYwEABCyADKMFDAUgDm5q48odj4RMRRUOYAjEkATOmZOcgAcj7+DRUh6GE1yJHRcYlwLUmFEsjtWQCiAI4SwoIjJErahcUlcAA+cN5+AYdwvhAQgsj6FgNAA

**Comment by swastik** [comment]
Excellent, thank you! I verified that works! 🙇🏼
